File: src/ios/UWebView.m
#import "UWebView.h"
#import "UWebViewRemoteVideoPlayer.h"
#import <UIKit/UIKit.h>

@interface UWebView ()

@property (nonatomic, strong) UWebViewRemoteVideoPlayer *remoteVideoPlayer;
@property (nonatomic, strong) CDVInvokedUrlCommand *callbackCommand;

@end

@implementation UWebView

- (void)pluginInitialize {
    NSLog(@"UWebView: initialization");
    [super pluginInitialize];
}

// 加载GeckoView URL
- (void)loadUrlWithGeckoView:(CDVInvokedUrlCommand *)command {
    NSString *url = [command.arguments objectAtIndex:0];
    [self openGeckoViewActivity:url];
}

// 初始化视频播放器
- (void)initVideoPlayer:(CDVInvokedUrlCommand *)command {
    [self playRemoteVideo:YES command:command];
}

// 播放远程视频
- (void)playRemoteVideo:(CDVInvokedUrlCommand *)command {
    [self playRemoteVideo:NO command:command];
}

// 关闭播放器
- (void)closePlayer:(CDVInvokedUrlCommand *)command {
    if (self.remoteVideoPlayer) {
        [self.remoteVideoPlayer closePlayer];
    }
}

// 销毁远程视频播放器
- (void)destroyRemoteVideo:(CDVInvokedUrlCommand *)command {
    if (self.remoteVideoPlayer) {
        [self.remoteVideoPlayer destroy];
        self.remoteVideoPlayer = nil;
    }
}

// 私有方法：打开GeckoView
- (void)openGeckoViewActivity:(NSString *)url {
    NSLog(@"UWebView: openGeckoViewActivity");
    // 假设这是打开一个新的UIViewController
    UIViewController *geckoViewController = [[UIViewController alloc] init];
    [self.viewController presentViewController:geckoViewController animated:YES completion:nil];
}

// 私有方法：播放远程视频
- (void)playRemoteVideo:(BOOL)isInit command:(CDVInvokedUrlCommand *)command {
    @try {
        NSString *url = [command.arguments objectAtIndex:0];
        CGFloat width = [[command.arguments objectAtIndex:1] floatValue];
        CGFloat height = [[command.arguments objectAtIndex:2] floatValue];
        CGFloat x = [[command.arguments objectAtIndex:3] floatValue];
        CGFloat y = [[command.arguments objectAtIndex:4] floatValue];
        
        if (width == 0) {
            width = CGRectGetWidth(self.viewController.view.frame);
        }

        if (height == 0) {
            height = width * (CGRectGetHeight(self.viewController.view.frame) / CGRectGetWidth(self.viewController.view.frame));
        }

        if (self.remoteVideoPlayer == nil) {
            dispatch_async(dispatch_get_main_queue(), ^{
                self.remoteVideoPlayer = [[UWebViewRemoteVideoPlayer alloc] initWithViewController:self.viewController
                                                                                              width:width
                                                                                             height:height
                                                                                                  x:x
                                                                                                  y:y];
                if (!isInit) {
                    [self.remoteVideoPlayer play:url];
                }
            });
        } else {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.remoteVideoPlayer updateLayoutWithWidth:width height:height x:x y:y];
                if (!isInit) {
                    [self.remoteVideoPlayer play:url];
                }
            });
        }
    } @catch (NSException *exception) {
        NSLog(@"Error: %@", exception.reason);
    }
}

@end

File: src/ios/UWebView.swift
import Foundation
import UIKit

@objc(UWebView)
public class UWebView: CDVPlugin {
    private static let TAG = "UWebView"
    public var callback: CDVInvokedUrlCommand?
    private var remoteVideoPlayer: GeckoViewRemoteVideoPlayer?

    @objc(initVideoPlayer:) 
    func initVideoPlayer(_ command: CDVInvokedUrlCommand) {
        playRemoteVideo2(isInit: true, command: command)  // 设置 `isInit: true`
    }

    @objc(playRemoteVideo:) 
    func playRemoteVideo(_ command: CDVInvokedUrlCommand) {
        playRemoteVideo2(isInit: false, command: command)
    }

    private func playRemoteVideo2(isInit: Bool, command: CDVInvokedUrlCommand) {
        do {
            let url = command.argument(at: 0) as! String
            let metrics = UIScreen.main.bounds.size
            let width = command.argument(at: 1) as! Int
            let height = command.argument(at: 2) as! Int
            let x = command.argument(at: 3) as! Int
            let y = command.argument(at: 4) as! Int
            
            let finalWidth = width == 0 ? Int(metrics.width) : width
            let finalHeight = height == 0 ? Int(Double(finalWidth) * (metrics.height / metrics.width)) : height

            DispatchQueue.main.async {
                if self.remoteVideoPlayer == nil {
                    self.remoteVideoPlayer = GeckoViewRemoteVideoPlayer(cordova: self.cordova, width: finalWidth, height: finalHeight, x: x, y: y)
                    if !isInit {
                        self.remoteVideoPlayer?.play(url: url)
                    }
                } else {
                    self.remoteVideoPlayer?.updateLayout(width: finalWidth, height: finalHeight, x: x, y: y)
                    if !isInit {
                        self.remoteVideoPlayer?.play(url: url)
                    }
                }
            }
        } catch {
            NSLog("Error in playRemoteVideo: \(error)")
        }
    }

    @objc(destroyVideoPlayer:)  
    func destroyVideoPlayer(_ command: CDVInvokedUrlCommand) -> Bool {
        if let player = remoteVideoPlayer {
            player.destroy()
            remoteVideoPlayer = nil
        }
        return true
    }

    @objc(destroyRemoteVideo:)  
    func destroyRemoteVideo(_ command: CDVInvokedUrlCommand) -> Bool {
        if let player = remoteVideoPlayer {
            player.destroy()
            remoteVideoPlayer = nil
        }
        return true
    }

    @objc(closePlayer:) 
    func closePlayer(_ command: CDVInvokedUrlCommand) -> Bool {
        remoteVideoPlayer?.closePlayer()
        return true
    }
}
